cpp/src/stream_compaction/distinct_count.cu

```diff
@@ -127,40 +129,39 @@ cudf::size_type distinct_count(table_view const& keys,
                                null_equality nulls_equal,
                                rmm::cuda_stream_view stream)
 {
-  auto const num_rows  = keys.num_rows();
+  auto const num_rows = keys.num_rows();
+  if (num_rows == 0) { return 0; }  // early exit for empty input
   auto const has_nulls = nullate::DYNAMIC{cudf::has_nested_nulls(keys)};
 
-  hash_map_type key_map{compute_hash_table_size(num_rows),
-                        cuco::empty_key{COMPACTION_EMPTY_KEY_SENTINEL},
-                        cuco::empty_value{COMPACTION_EMPTY_VALUE_SENTINEL},
-                        detail::hash_table_allocator_type{default_allocator<char>{}, stream},
-                        stream.value()};
-
   auto const preprocessed_input =
     cudf::experimental::row::hash::preprocessed_table::create(keys, stream);
-
   auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
   auto const hash_key   = experimental::compaction_hash(row_hasher.device_hasher(has_nulls));
-
-  auto const row_comp = cudf::experimental::row::equality::self_comparator(preprocessed_input);
-
-  auto iter = cudf::detail::make_counting_transform_iterator(
-    0, [] __device__(size_type i) { return cuco::make_pair(i, i); });
+  auto const row_comp   = cudf::experimental::row::equality::self_comparator(preprocessed_input);
 
   auto const comparator_helper = [&](auto const row_equal) {
+    using hasher_type = decltype(hash_key);
+    auto key_set      = cuco::experimental::static_set{
+      cuco::experimental::extent{static_cast<cudf::size_type>(compute_hash_table_size(num_rows))},
```

Comment:
Is there any guarantee that `compute_hash_table_size` will be within integer limit?

Reply:
Answer to myself: it is not. So the output of `compute_hash_table_size` is around `1.5 * num_rows` which may overflow here.

Reply:
I will add a runtime dispatching to this

Reply:
Did a test on this and `extent<size_type>` wouldn't bring any speedups over `extent<std::size_t>`. Thus no more casting.